extensions/podman/packages/extension/src/podman-machine-stream.ts

```diff
@@ -0,0 +1,147 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import * as fs from 'node:fs';
+
+import type {
+  Event,
+  ProviderConnectionShellAccess,
+  ProviderConnectionShellAccessData,
+  ProviderConnectionShellAccessError,
+  ProviderConnectionShellAccessSession,
+  ProviderConnectionShellDimensions,
+} from '@podman-desktop/api';
+import { Disposable, EventEmitter } from '@podman-desktop/api';
+import type { ClientChannel } from 'ssh2';
+import { Client } from 'ssh2';
+
+import type { MachineInfo } from './extension';
+
+export class ProviderConnectionShellAccessImpl extends Disposable implements ProviderConnectionShellAccess {
+  #host: string;
+  #port: number;
+  #username: string;
+  #privateKey: string;
+  #client: Client | undefined;
+  #stream: ClientChannel | undefined;
+  #connected: boolean = false;
+
+  constructor(private readonly podmanMachine: MachineInfo) {
+    super(Disposable);
+    this.#host = 'localhost';
+    this.#port = podmanMachine.port;
+    this.#username = podmanMachine.remoteUsername;
+    this.#privateKey = podmanMachine.identityPath;
+  }
+
+  dispose(): void {
+    this.close();
+  }
+
+  protected isConnected(): boolean {
+    return this.#connected;
+  }
+
+  onDataEmit = new EventEmitter<ProviderConnectionShellAccessData>();
+  onData: Event<ProviderConnectionShellAccessData> = this.onDataEmit.event;
+
+  onErrorEmit = new EventEmitter<ProviderConnectionShellAccessError>();
+  onError: Event<ProviderConnectionShellAccessError> = this.onErrorEmit.event;
+
+  onEndEmit = new EventEmitter<void>();
+  onEnd: Event<void> = this.onEndEmit.event;
+
+  write(data: string): void {
+    this.#stream?.write(data);
+  }
+
+  resize(dimensions: ProviderConnectionShellDimensions): void {
+    // rows and cols override width and height when rows and cols are non-zero.
+    this.#stream?.setWindow(dimensions.rows, dimensions.cols, 0, 0);
+  }
+
+  protected disposeListeners(): void {
+    this.onDataEmit.dispose();
+    this.onErrorEmit.dispose();
+    this.onEndEmit.dispose();
+  }
+
+  close(): void {
+    this.#connected = false;
+    this.#client?.end();
+    this.#client?.destroy();
+    this.disposeListeners();
+  }
+
+  open(): ProviderConnectionShellAccessSession {
+    // Avoid multiple connections
+    if (this.#connected) {
+      console.warn('Removing listeners');
+      this.disposeListeners();
```

Comment:
please remove the console warn

about the if condition, could you please tell in which case does it happen . I think this is the part where I'm not sure to follow and why there is a specific disposeListeners method etc

So it makes me thing that there is something to dig as I have the feeling here that if I request, as Podman Desktop, to open two connections, I will only be connected to a single ssh session 


IMHO, each open should return a new independent session. so we should not have the 'if connected'

Reply:
I suggested the warn as it seems an edge case that should not happen

Reply:
The open is called every time when you click on a "terminal," and this function will be called https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/provider-registry.ts#L1284, which calls "open" (opens a new session and adds listeners for it).

But when you go to a different tab and go back to the "terminal" +1 listener is added 
This results in duplicated data in the terminal; that's why there is the `if`, which checks if a given providerConnection already exists session, and if yes, removes all listeners (because they are added in lines L1284+ anyways). 

Another edge case is typing "exit" which also some problems executing 
There should be all the changes / "working solution" but without suggestions from reviews https://github.com/gastoner/podman-desktop/tree/provide_ability_to_easily_get_a_shell_in_a_machine_test

Reply:
I expect to have a new session each time I call open

so there should not be any isConnected stuff

how the UI will reuse a session or create a new one should not be up to the extension

The API is saying that calling open returns a new session, so I expect to have a dedicated session/terminal on each open call, not to share with others my session